python/tank/platform/software_launcher.py

```diff
@@ -0,0 +1,406 @@
+        This can be useful if you want to write app code
+        to retrieve a local resource::
+
+            app_font = os.path.join(self.disk_location, "resources", "font.fnt")
+        """
+        path_to_this_file = os.path.abspath(sys.modules[self.__module__].__file__)
+        return os.path.dirname(path_to_this_file)
+
+    @property
+    def display_name(self):
+        """
+        The display name for the item. Automatically
+        appends 'Startup' to the end of the display
+        name if that string is missing from the display
+        name (e.g. Maya Engine Startup)
+
+        :returns: display name as string
+        """
+        disp_name = self.descriptor.display_name
+        if not disp_name.lower().endswith("startup"):
+            # Append "Startup" to the default descriptor
+            # display_name to distinguish it from the
+            # engine's display name.
+            disp_name = "%s Startup" % disp_name
+        return disp_name
+
+    @property
+    def engine_name(self):
+        """
+        The TK engine name this launcher is based on.
+
+        :returns: String TK engine name
+        """
+        return self.__engine_name
+
+    @property
+    def logger(self):
+        """
+        Standard python logger for this engine, app or framework.
+        Use this whenever you want to emit or process log messages.
+
+        :returns: logging.Logger instance
+        """
+        return LogManager.get_logger("env.%s.%s.startup" %
+            (self.__environment.name, self.__engine_name)
+        )
+
+    @property
+    def shotgun(self):
+        """
+        Returns a Shotgun API handle associated with the currently running
+        environment. This method is a convenience method that calls out
+        to :meth:`~sgtk.Tank.shotgun`.
+
+        :returns: Shotgun API handle
+        """
+        return self.sgtk.shotgun
+
+
+    ##########################################################################################
+    # abstract methods
+
+    def scan_software(self, versions=None, display_name=None, icon=None):
+        """
+        Performs a scan for software installations.
+
+        :param versions: List of strings representing versions
+                         to search for. If set to None, search
+                         for all versions. A version string is
+                         DCC-specific but could be something
+                         like "2017", "6.3v7" or "1.2.3.52"
+        :param display_name : (optional) String to use to describe the
+                              resulting SoftwareVersion(s) in graphical
+                              displays.
+        :param icon: (optional) Path to icon to use with the resulting
+                     SoftwareVersion(s) in graphical displays.
+        :returns: List of :class:`SoftwareVersion` instances
+        """
+        raise NotImplementedError
+
+    def prepare_launch(self, exec_path, args, options, file_to_open=None):
```

Comment:
I think we should get rid of `options` for the time being. There isn't a clear use case and we can always add it later if such a use case emerges. What do you think? It was in the original designs to support the case with photoshop, but I realise now (seeing your launchapp code) that the better solution in that case would simply be to add some engine config parameters rather than passing those parameters from the launch app into the engine. Sorry about the back and forth on this!